### PR TITLE
Save lora config

### PR DESCRIPTION
**Pull request by awni**
- Save lora experiment config to json config
- Use safetensors for adapters
-  Read config as well as weights when loading lora model

Directory structure like:

```
adapters/
    adapters.safetensors
    adapter_config.json
```

Config looks like this:

```json
{
    "adapter_path": "adapters",
    "batch_size": 4,
    "config": null,
    "data": "../lora/data/",
    "grad_checkpoint": false,
    "iters": 11,
    "learning_rate": 1e-05,
    "lora_layers": 16,
    "lora_parameters": {
        "rank": 8,
        "alpha": 16,
        "dropout": 0.0,
        "scale": 10.0
    },
    "max_seq_length": 2048,
    "model": "Qwen/Qwen1.5-0.5B",
    "resume_adapter_file": null,
    "save_every": 10,
    "seed": 0,
    "steps_per_eval": 200,
    "steps_per_report": 10,
    "test": true,
    "test_batches": 500,
    "train": true,
    "val_batches": 25
}
```

### CODE REVIEW

**Comment by awni** [comment]
@mzbac would you mind looking at this.

File: llms/mlx_lm/tuner/trainer.py [hunk 1]
```diff
@@ -287,13 +276,10 @@ def step(batch):
...
 
-        # Save adapter weights if needed
+        # Save adapter weights
         if (it + 1) % args.steps_per_save == 0:
-            checkpoint_adapter_file = (
-                f"{adapter_path}/{it + 1}_{Path(args.adapter_file).name}"
-            )
-            save_adapter(model=model, adapter_file=checkpoint_adapter_file)
-            print(f"Iter {it + 1}: Saved adapter weights to {checkpoint_adapter_file}.")
+            save_adapter(model=model, adapter_file=args.adapter_file)
```
Comment (mzbac) [inline llms/mlx_lm/tuner/trainer.py:281]: Is this going to keep overriding the same adapter file, causing us to lose all the intermediate checkpoint weights?

**Comment by mzbac** [comment]
Looks good, just a question about how to handle the weights for the checkout point. Also, we may need to update the server.py script to use adapter-path instead of adapter-file.

File: llms/mlx_lm/tuner/trainer.py (hunk 1 above)
Comment (mzbac) [reply llms/mlx_lm/tuner/trainer.py:281]: I am thinking maybe we can do something like this:
```
├── lora_adapters/
│   ├── adapter_1/
│   │   ├── adapter_config.json  
│   │   └── adapters.safetensors
│   ├── adapter_2/
│   │   ├── adapter_config.json    
│   │   └── adapters.safetensors
│   ├── adapter_N/
│   |   ├── adapter_config.json    
│   |   └── adapters.safetensors
│   └── adapter_final/
│       ├── adapter_config.json    
│       └── adapters.safetensors
```
It may seem a bit redundant with the `adapter_config.json`, but we would save more precise information for that training iteration. For example, the learning rate of the scheduled learning rate, loss, and anything else related to the training iteration.

Comment (awni) [reply llms/mlx_lm/tuner/trainer.py:281]: I went with
```
lora_adapters/
  -adapters.safetensors
  -adapter_config.json
  -00100_adapters.safetensors
...
``

That fits the way we load better from resumed files and saves from needing to pass all the args into the traiing file or hacking a way around copying the config.

**Comment by awni** [comment]
Comments addressed, thanks @mzbac !

**Comment by awni** [comment]
Closes #622

**Review by angeloskath — APPROVED** [review]

File: llms/mlx_lm/tuner/trainer.py (hunk 1 above)
Comment (mzbac) [reply llms/mlx_lm/tuner/trainer.py]: Alright, maybe we could add an optional adapter file name parameter to allow users to specify a specific checkpoint weight during the inference so they don't have to copy the checkpoint weight into a folder and rename it in order to load it. If you are okay with this approach, I can work on a PR for adding that support.

Comment (awni) [reply llms/mlx_lm/tuner/trainer.py]: I'm ok with that.. I just didn't think it was a common thing people did?

Comment (awni) [reply llms/mlx_lm/tuner/trainer.py]: Do you do it yourself? If so, what for?

Comment (mzbac) [reply llms/mlx_lm/tuner/trainer.py]: Yeah, maybe not common use cases, but something from my personal experience. It's not always the adapter with the lowest loss that performs the best, so I normally test out a few adapters to see which one is the best. However, I'm happy to just wait and see if this is something people want to add.

Comment (awni) [reply llms/mlx_lm/tuner/trainer.py]: Do you test them on a downstream task? I wonder if we saved the adapter based on validation loss maybe that would help.
